python/social_bot/envs/embodied_teacher.py

```diff
@@ -0,0 +1,268 @@
+                step_time is rounded to multiples of world_time_precision
+                step_time / world_time_precision is how many simulator substeps during one
+                environment step. for the tasks need higher control frequency (such as the 
+                tasks need walking by 2 legs), using a smaller step_time like 0.05 is better.
+                experiments show that iCub can not learn how to walk in a 0.1 step_time
+            real_time_update_rate (int): max update_rate per seconds. there is no limit if
+                this is set to 0. if 1:1 real time is prefered(like playing or recording video),
+                this should be set to 1.0/world_time_precision.
+            port: Gazebo port, need to specify when run multiple environment in parallel
+            action_cost (float): Add an extra action cost to reward, which helps to train
+                an energy/forces efficency policy or reduce unnecessary movements
+            resized_image_size (None|tuple): If None, use the original image size
+                from the camera. Otherwise, the original image will be resized
+                to (width, height)
+            vocab_sequence_length (int): the length of encoded sequence
+            initial_teacher_pose (string): initial teacher pose in the world, the format
+                is "x y z roll pitch yaw"
+            demo_by_human (bool): demo by human or by existing teacher policy
+        """
+        self._demo_by_human = demo_by_human
+        if self._demo_by_human:
+            from social_bot.keybo_control import KeyboardControl
+            self._keybo = KeyboardControl()
+            real_time_update_rate = 500  # run "gz physics -u" to override
+        
+        super().__init__(
+            agent_type=agent_type,
+            world_name="play_ground.world",
+            tasks=tasks,
+            with_language=with_language,
+            use_image_observation=use_image_observation,
+            image_with_internal_states=image_with_internal_states,
+            world_time_precision=world_time_precision,
+            step_time=step_time,
+            real_time_update_rate=real_time_update_rate,
+            port=port,
+            action_cost=action_cost,
+            resized_image_size=resized_image_size,
+            vocab_sequence_length=vocab_sequence_length)
+
+        # insert teacher model
+        self.insert_model(
+            model=agent_type, name="teacher", pose=initial_teacher_pose)
+        # set up teacher joints
+        self._teacher_joints = []
+        for joint in self._agent_joints:
+            self._teacher_joints.append("teacher::" + joint)
+        self._teacher_embodied = self._world.get_agent('teacher')
+        
+        # setup action and observation space
+        if not self._demo_by_human:
+            self.action_space = gym.spaces.Dict(
+                learner=self.action_space,
```

Comment:
Teacher and leaner might have different action spaces, e.g., in the case of different morphologies. 
From Line186-Line187, it seems that it is supports different actions spaces for teacher and learner.  But from Line 132 to Line 133, it is hard coded to the same action space.
This is something for the future can could be done later.

Reply:
Yes, currently teacher's action and observation are hard coded to the same space. Since it depends on the configurations of environment where existing teacher's policy are trained, perhaps we should make the teacher's space configurable?

Reply:
We can keep it as is and leave this to a future PR.